### PR TITLE
Custom timeline track and clip for cutscene dialogue

### PR DESCRIPTION
**Pull request by Caiuse**
card: https://open.codecks.io/unity-open-project-1/decks/15/card/15r-cutscene-system

- Custom timeline track and clip
- Each clip has settings which using the `dialogueID` option as a ID for the line of dialogue needed to be displayed
- Each track can has a binder assigned which can later be inherited to build the logic for binding text to the screen.
- If `stopGraphOnClipEnd` clip option is selected the graph will pause and wait for another system to unpause the graph.

**Workflow**

Cutscene designers creates a new dialogue track within the main timeline, then for each line of dialog a new clip is created and assigned a ID which will later be used to display ( or play audio of ) the corresponding dialogue, the clip can then be set to pause the timeline when it reaches its end time, and another system can be used to unpause and continue the cutscene.

Screen Recording 2020-10-13 at 19 06 26

### CODE REVIEW

**Review by dkaloger — CHANGES_REQUESTED** [review]
overlap with https://github.com/UnityTechnologies/open-project-1/pull/56

**Comment by Caiuse** [comment]
Used @Aspekt1024 dialog track as the base on this PR and adapted the functionality after talking with him. 

@dkaloger reviewed your PR, once those bits have been addressed we can raise another PR which merges your rendering. These two systems can remain decoupled for the moment making Unity's review a little easier 🤞

**Review by 0479010347 — APPROVED** [review]

**Comment by 0479010347** [comment]
What is this

**Comment by ardaozler** [comment]
I didnt mean to issue a pull req. I must have done it by accident. Sorry! 

Android’de Yahoo Postadan gönderildi 
 
  15:47’’16e’ 16 Eki 2020 Cum tarihinde, 0479010347<notifications@github.com> şunu yazdı:   


What is this

**Comment by ardaozler** [comment]
I just realized that i recieved this because im on watch sorry again lol.

Android’de Yahoo Postadan gönderildi 
 
  15:55’’16e’ 16 Eki 2020 Cum tarihinde, Arda Özler<ardaozler@yahoo.com> şunu yazdı:   I didnt mean to issue a pull req. I must have done it by accident. Sorry! 

Android’de Yahoo Postadan gönderildi 
 
  15:47’’16e’ 16 Eki 2020 Cum tarihinde, 0479010347<notifications@github.com> şunu yazdı:   


What is this

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+csharp_space_before_dot = false
+csharp_space_after_semicolon_in_for_statement = true
+csharp_space_before_semicolon_in_for_statement = false
+csharp_space_around_declaration_statements = false
+csharp_space_before_open_square_brackets = false
+csharp_space_between_empty_square_brackets = false
+csharp_space_between_square_brackets = false
+csharp_preserve_single_line_statements = false
+csharp_preserve_single_line_blocks = true
+csharp_using_directive_placement = outside_namespace
```
Comment (ciro-unity) [inline .editorconfig:38]: Why is this file part of the PR?

File: UOP1_Project/Assets/Scripts/Cutscenes/DialogueTimeline/DialogueMixerBehaviour.cs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+			for (int i = 0; i < m_clipsData.Length; i++)
+			{
+				if (currentTime >= m_clipsData[i].Range.Start && currentTime < m_clipsData[i].Range.End)
+				{
+					m_currentClipIndex = i;
+					m_clipsData[m_currentClipIndex].state = ClipState.Pending;
+					break;
+				}
+			}
+		}
```
Comment (ciro-unity) [inline UOP1_Project/Assets/Scripts/Cutscenes/DialogueTimeline/DialogueMixerBehaviour.cs:64]: Are you looking just to know which clip is currently played by the Timeline? You should use weights.

@rainaldisatria is doing it here for instance: https://github.com/UnityTechnologies/open-project-1/blob/c1aca3610d73fa83acb43e72e895f349c4250fbf/UOP1_Project/Assets/Scripts/Cutscenes/Dialogue%20Timeline/DialogueControlMix.cs

If the weight is 0, the clip is not being played. If it's 1, it's played and it's the only one.
More than one clip can be played at a time due to a blend. In which case the weight won't be 1 but something in between.

File: UOP1_Project/Assets/Scripts/Cutscenes/DialogueTimeline/DialogueBinder.cs [hunk 3]
```diff
@@ -0,0 +1,11 @@
+
+namespace UOP1.Cutscenes
+{
+	/// <summary>
+	/// Inherit this class to create a target for the timelines dialogue
+	/// </summary>
+	public abstract class DialogueBinder : UnityEngine.Component
+	{
+		public abstract void SetDialog(string dialogueLineID);
```
Comment (ciro-unity) [inline UOP1_Project/Assets/Scripts/Cutscenes/DialogueTimeline/DialogueBinder.cs:9]: The idea of calling a dialogue line from a ScriptableObject using a string sounds like a dangerous idea, it's prone to mistakes and doesn't work well when you decide to move lines around. Now you need to manually go and look into clips to change the IDs.

**Review by ciro-unity — CHANGES_REQUESTED** [review]
A nice start and thanks for making it short, but there are some things I'd like to fix before pulling it.

**Comment by ciro-unity** [comment]
Closing because in the end I merged PR #135, with some code from this one too.

**Pull request closed without merging**